programs/vote/src/vote_state.rs

```diff
@@ -489,6 +564,55 @@ pub fn process_vote(
     vote_account.set_state(&vote_state)
 }
 
+pub fn slash_state(
+    vote_account: &mut KeyedAccount,
+    slot_hashes: &[SlotHash],
+    program_id: &Pubkey,
+    tx: &Transaction,
+) -> Result<(), InstructionError> {
+    let mut vote_state: VoteState = vote_account.state()?;
+    //verify the tx is valid
+    if !tx.verify_refs() {
+        return Err(InstructionError::CustomError(
+            VoteError::InvalidSlashTransaction as u32,
+        ));
+    }
+    //verify transaction signature
+    if tx.verify().is_err() {
+        return Err(InstructionError::CustomError(
+            VoteError::InvalidSlashTransaction as u32,
+        ));
+    }
+    let signers: HashSet<Pubkey> = tx
+        .message
+        .account_keys
+        .iter()
+        .take(tx.signatures.len())
+        .cloned()
+        .collect();
+    //find vote instruction
+    for (i, ix) in tx.message.instructions.iter().enumerate() {
+        if tx.instruction_program(i) != Some(&program_id) {
+            continue;
+        }
+        if let Ok(VoteInstruction::Vote(vote)) = limited_deserialize(&ix.data) {
+            //verify that account matches mine
+            if tx.key(i, 0) != Some(vote_account.unsigned_key()) {
+                continue;
+            }
+            //verify that the transaction was signed by the expected signers
+            if verify_authorized_signer(&vote_state.authorized_voter, &signers).is_err() {
```

Comment:
there's now a history of vote signers in vote_state that we should also be checking against